### PR TITLE
Add estimate fee endpoint

### PR DESCRIPTION
**Pull request by paweljakubas**
# Issue Number

#469 

# Overview

- [x] I have added implementation of estimateFee endpoint
- [x] I have added respective typesSpec unit tests
- [x] I have added corresponding integration tests unifying them with already present ones

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet.hs [hunk 1]
```diff
@@ -607,6 +626,23 @@ newWalletLayer tracer block0 feePolicy db nw tl = do
...
+        => WalletId
+        -> CoinSelectionOptions
+        -> NonEmpty TxOut
+        -> ExceptT ErrEstimateTxFee IO Fee
+    _estimateTxFee wid opts recipients = do
+        (w, _) <- withExceptT ErrEstimateTxFeeNoSuchWallet (_readWallet wid)
+        let utxo = availableUTxO @s @t w
+        (sel, _utxo') <- withExceptT ErrEstimateTxFeeCoinSelection $
+            CoinSelection.random opts recipients utxo
+        logInfoT $ "Coins selected for transaction: \n"+| sel |+""
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: I think this log line is superfluous in `estimateTxFee`.

File: specifications/api/swagger.yaml [hunk 2]
```diff
@@ -952,11 +959,11 @@ paths:
       tags: ["Transactions"]
       summary: Estimate
       description: |
-        <p align="right">status: <strong>not implemented</strong></p>
+        <p align="right">status: <strong>stable</strong></p>
```
Comment (KtorZ) [inline specifications/api/swagger.yaml:962]: :tada:

File: lib/core/src/Cardano/Wallet.hs [hunk 3]
```diff
@@ -278,6 +289,13 @@ data ErrCreateUnsignedTx
     | ErrCreateUnsignedTxFee ErrAdjustForFee
     deriving (Show, Eq)
 
+-- | Errors occuring when estimating transaction fee
+data ErrEstimateTxFee
+    = ErrEstimateTxFeeNoSuchWallet ErrNoSuchWallet
+    | ErrEstimateTxFeeCoinSelection ErrCoinSelection
+    | ErrEstimateTxFeeFee ErrAdjustForFee
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: Is this last branch really reachable in `estimateTxFee` ? From the top of my head `ErrAdjustForFee` is only returned by `adjustForFee` which isn't called in the fee estimation. Right?

**Review by KtorZ — COMMENTED** [review]
Awesome work @paweljakubas !
I haven't look in depth to the integration scenarios. From a quick glance, this looks okay, but I'll spend a bit more time later reviewing carefully and I invite @piotr-iohk to also do the same :smile:

File: lib/core/src/Cardano/Wallet.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/core/src/Cardano/Wallet.hs]: correct!  it is not needed as we do not use `adjustForFees` here

File: lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs [hunk 4]
```diff
@@ -734,3 +631,510 @@ spec = do
...
+                    "address": #{destination},
+                    "amount": {
+                        "quantity": 1,
+                        "unit": "lovelace"
+                    }
+                }]
+            }|]
+        r <- request @ApiFee ctx (postTxFeeEp wSrc) Default payload
+        verify r
+            [ expectResponseCode HTTP.status202 ]
```
Comment (piotr-iohk) [inline lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs:851]: perhaps let's also check the amount here: `expectFieldBetween amount...`

**Review by piotr-iohk — COMMENTED** [review]
I've reviewed the tests - looking good, thanks a lot @paweljakubas ! :)
There's one test that could be added: estimate transaction for amount = `0`. That one should behave different for `http-bridge` and `jormungandr` so they should go to their specific libs, like:

`http-bridge` -> https://github.com/input-output-hk/cardano-wallet/blob/master/lib/http-bridge/test/integration/Test/Integration/HttpBridge/Scenario/API/Transactions.hs

`jormungandr` -> https://github.com/input-output-hk/cardano-wallet/tree/master/lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/

File: lib/core/src/Cardano/Wallet.hs [hunk 5]
```diff
@@ -278,6 +289,12 @@ data ErrCreateUnsignedTx
     | ErrCreateUnsignedTxFee ErrAdjustForFee
     deriving (Show, Eq)
 
+-- | Errors occuring when estimating transaction fee
+data ErrEstimateTxFee
+    = ErrEstimateTxFeeNoSuchWallet ErrNoSuchWallet
+    | ErrEstimateTxFeeCoinSelection ErrCoinSelection
+    deriving (Show, Eq)
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:296]: :+1:

**Review by KtorZ — APPROVED** [review]

**Pull request merged**